synapseclient/core/pool_provider.py

```diff
@@ -50,13 +67,29 @@ def __init__(self, type, value):
     def get_lock(self):
         return FakeLock()
 
+
 def get_pool():
     if config.single_threaded:
         return SingleThreadPool()
     else:
         return multiprocessing.dummy.Pool(DEFAULT_POOL_SIZE)
 
 
+def get_executor(thread_count=DEFAULT_POOL_SIZE):
+    """
+    Provides an Executor as defined by the client config suitable
+    for running tasks work as defined by the client config.
+
+    :param thread_count:        number of concurrent threads
+
+    :return: an Executor
+    """
+    if config.single_threaded:
```

Comment:
we should get rid of the config and just attempt to detect if threading is possible.
I'm not sure if ThreadPoolExecutor handles dummy_threading or not already, but if it does then we have don't have to do anything. https://docs.python.org/3.6/library/dummy_threading.html

Reply:
I agree the config setting is currently strange. It's documented here thoughhttps://python-docs.synapse.org/build/html/news.html?highlight=single_thread#id18, so I'm thinking at least one release to announce it's being removed and maybe some stderr warning when setting it would be nice rather than removing it.

I think a user exposed option to set the number of threads specifically i.e. aws client's max_concurrent_requests would be good.